src/subject.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+trait Subject {
+    fn is_valid(&self) -> bool;
+    fn tokens(&self) -> std::io::Result<Vec<&str>>;
+    fn token_at(&self, n: usize) -> std::io::Result<&str>;
+}
+
+impl Subject for &str {
+    fn is_valid(&self) -> bool {
+        let mut fwc_seen = false;
+        for token in self.split('.') {
+            if !token_valid(token) || fwc_seen {
+                return false;
+            }
+            fwc_seen = token_is_fwc(token);
+        }
+        true
+    }
+
+    fn tokens(&self) -> std::io::Result<Vec<&str>> {
+        let tokens = self.split('.').collect::<Vec<&str>>();
+        let mut fwc_seen = false;
+        for (i, token) in tokens.iter().enumerate() {
+            if !token_valid(token) || fwc_seen {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::InvalidInput,
+                    format!("invalid token at position {}", i),
+                ));
+            }
+            fwc_seen = token_is_fwc(token);
+        }
+        Ok(tokens)
+    }
+
+    fn token_at(&self, n: usize) -> std::io::Result<&str> {
+        let tokens = self.tokens()?;
```

Comment:
Here you pay for the design of `Subject::tokens()`. You allocate a `Vec` only to look up one token.

Reply:
Agree, this was just playing around a bit. Nothing too serious here, just spinning on your concepts a bit for a lighter weight version API wise (not impl).